libsplinter/src/admin/client/event/ws.rs

```diff
@@ -0,0 +1,439 @@
+            }
+        }
+    }
+
+    fn wait_for_shutdown(mut self) -> Result<(), InternalError> {
+        match &mut self.ws_runtime {
+            WsRuntime::Reactor(reactor) => {
+                if let Some(reactor) = reactor.take() {
+                    reactor
+                        .wait_for_shutdown()
+                        .map_err(|e| InternalError::from_source(Box::new(e)))
+                } else {
+                    // Calling this function will have consumed this object, so we don't have any
+                    // alternative branches
+                    unreachable!()
+                }
+            }
+            _ => Ok(()),
+        }
+    }
+}
+
+impl Drop for WsAdminServiceEventClient {
+    fn drop(&mut self) {
+        self.signal_shutdown();
+    }
+}
+
+impl AdminServiceEventClient for WsAdminServiceEventClient {
+    /// Non-blocking
+    fn try_next_event(&self) -> Result<AdminServiceEvent, TryNextEventError> {
+        let evt_result = self.event_receiver.try_recv().map_err(|e| match e {
+            TryRecvError::Empty => TryNextEventError::Empty,
+            TryRecvError::Disconnected => TryNextEventError::Disconnected,
+        })?;
+
+        evt_result
+            .map_err(|e| TryNextEventError::InternalError(InternalError::from_source(Box::new(e))))
+    }
+
+    /// Blocking
+    fn next_event(&self) -> Result<AdminServiceEvent, NextEventError> {
+        let evt_result = self
+            .event_receiver
+            .recv()
+            .map_err(|_| NextEventError::Disconnected)?;
+        evt_result
+            .map_err(|e| NextEventError::InternalError(InternalError::from_source(Box::new(e))))
+    }
+}
+
+impl ParseBytes<AdminServiceEvent> for AdminServiceEvent {
+    fn from_bytes(bytes: &[u8]) -> Result<AdminServiceEvent, ParseError> {
+        let json_event: Event = serde_json::from_slice(bytes)
+            .map_err(|err| ParseError::MalformedMessage(Box::new(err)))?;
+
+        use messages::AdminServiceEvent::*;
+        let (proposal, event_type) = match json_event.admin_event {
+            ProposalSubmitted(proposal) => (proposal, EventType::ProposalSubmitted),
+            ProposalVote((proposal, pub_key_bytes)) => (
+                proposal,
+                EventType::ProposalVote {
+                    requester: PublicKey(pub_key_bytes),
+                },
+            ),
+            ProposalAccepted((proposal, pub_key_bytes)) => (
+                proposal,
+                EventType::ProposalAccepted {
+                    requester: PublicKey(pub_key_bytes),
+                },
+            ),
+            ProposalRejected((proposal, pub_key_bytes)) => (
+                proposal,
+                EventType::ProposalRejected {
+                    requester: PublicKey(pub_key_bytes),
+                },
+            ),
+            CircuitReady(proposal) => (proposal, EventType::CircuitReady),
+            CircuitDisbanded(proposal) => (proposal, EventType::CircuitReady),
```

Comment:
This result should be `EventType::CircuitDisbanded`, which I believe was causing the error you ran into with the disband tests.

Reply:
Fixed